common/scala/src/main/scala/org/apache/openwhisk/core/yarn/YARNComponentActor.scala

```diff
@@ -0,0 +1,101 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.openwhisk.core.yarn
+
+import akka.actor.{Actor, ActorSystem}
+import akka.http.scaladsl.model.{HttpMethods, StatusCodes}
+import akka.stream.ActorMaterializer
+import org.apache.openwhisk.common.Logging
+import org.apache.openwhisk.core.entity.ExecManifest.ImageName
+import org.apache.openwhisk.core.yarn.YARNComponentActor.{CreateContainerAsync, RemoveContainer}
+
+import scala.concurrent.ExecutionContext
+
+//Submits create command to YARN and submits decommission command to YARN
+object YARNComponentActor {
+  case object CreateContainerAsync
+  case class RemoveContainer(component_instance_name: String)
+}
+class YARNComponentActor(actorSystem: ActorSystem,
+                         logging: Logging,
+                         yarnConfig: YARNConfig,
+                         serviceName: String,
+                         imageName: ImageName)
+    extends Actor {
+
+  implicit val as: ActorSystem = actorSystem
+  implicit val materializer: ActorMaterializer = ActorMaterializer()
+  implicit val ec: ExecutionContext = actorSystem.dispatcher
+
+  var containerCount: Int = 0
+
+  def receive: PartialFunction[Any, Unit] = {
+    case CreateContainerAsync =>
+      sender ! createContainerAsync
+
+    case RemoveContainer(component_instance_name) =>
+      sender ! removeContainer(component_instance_name)
+
+    case input =>
+      throw new IllegalArgumentException("Unknown input: " + input)
+      sender ! false
+  }
+
+  def createContainerAsync(): Unit = {
+
+    logging.info(this, s"Using YARN to create a container with image ${imageName.name}...")
+
+    val body = "{\"number_of_containers\":" + (containerCount + 1) + "}"
+    val response = YARNRESTUtil.submitRequestWithAuth(
+      yarnConfig.authType,
+      HttpMethods.PUT,
+      s"${yarnConfig.masterUrl}/app/v1/services/$serviceName/components/${imageName.name}",
+      body)
+    response match {
+      case httpresponse(StatusCodes.OK, content) =>
+        logging.info(this, s"Added container: ${imageName.name}. Response: $content")
+        containerCount += 1
+
+      case httpresponse(_, _) => YARNRESTUtil.handleYARNRESTError(logging)
+    }
+  }
+
+  def removeContainer(component_instance_name: String): Unit = {
+    logging.info(this, s"Removing ${imageName.name} container: $component_instance_name ")
+    if (containerCount <= 0) {
+      logging.warn(this, "Already at 0 containers")
+    } else {
+      val body =
+        s"""{"components":[ { "name":"${imageName.name}", "decommissioned_instances":["$component_instance_name"] }]}"""
```

Comment:
also can be done with JsObject - ok here since there are small objects and clear.

Reply:
I changed it to JsObject. It is a bit cleaner.